src/render_tree.rs

```diff
@@ -0,0 +1,563 @@
+            if let Some(current_node) = doc_read.get_node_by_id(current_node_id) {
+                match &current_node.data {
+                    NodeData::Element(element) => {
+                        match element.name.as_str() {
+                            "h1" => {
+                                let new_node =
+                                    Rc::new(RefCell::new(Node::new_heading1(&mut self.position)));
+                                util::add_text_node(&reference_element, &new_node);
+                                reference_element = Rc::clone(&new_node);
+                            }
+                            "h2" => {
+                                let new_node =
+                                    Rc::new(RefCell::new(Node::new_heading2(&mut self.position)));
+                                util::add_text_node(&reference_element, &new_node);
+                                reference_element = Rc::clone(&new_node);
+                            }
+                            "h3" => {
+                                let new_node =
+                                    Rc::new(RefCell::new(Node::new_heading3(&mut self.position)));
+                                util::add_text_node(&reference_element, &new_node);
+                                reference_element = Rc::clone(&new_node);
+                            }
+                            "h4" => {
+                                let new_node =
+                                    Rc::new(RefCell::new(Node::new_heading4(&mut self.position)));
+                                util::add_text_node(&reference_element, &new_node);
+                                reference_element = Rc::clone(&new_node);
+                            }
+                            "h5" => {
+                                let new_node =
+                                    Rc::new(RefCell::new(Node::new_heading5(&mut self.position)));
+                                util::add_text_node(&reference_element, &new_node);
+                                reference_element = Rc::clone(&new_node);
+                            }
+                            "h6" => {
+                                let new_node =
+                                    Rc::new(RefCell::new(Node::new_heading6(&mut self.position)));
+                                util::add_text_node(&reference_element, &new_node);
+                                reference_element = Rc::clone(&new_node);
+                            }
+                            "p" => {
+                                let new_node =
+                                    Rc::new(RefCell::new(Node::new_paragraph(&mut self.position)));
+                                util::add_text_node(&reference_element, &new_node);
+                                reference_element = Rc::clone(&new_node);
+                            }
+                            _ => { /* add more here, for now ignore */ }
+                        }
```

Comment:
You have repeating code here, this can be improved significantly. I have something ready, I can commit it if you want.

Reply:
can you commit into a separate branch? that way if we decide not to go forward we doubt have to deal with reverting commits. Otherwise we could then commit into this branch and rebase